contracts/near/eth2-client/src/lib.rs

```diff
@@ -259,41 +213,73 @@ impl Eth2Client {
 
     #[result_serializer(borsh)]
     pub fn submit_execution_header(&mut self, #[serializer(borsh)] block_header: BlockHeader) {
-        if self.finalized_beacon_header.execution_block_hash != block_header.parent_hash {
-            self.unfinalized_headers
-                .get(&block_header.parent_hash)
-                .unwrap_or_else(|| {
-                    env::panic_str(
-                        format!(
-                            "Header has unknown parent {:?}. Parent should be submitted first.",
-                            block_header.parent_hash
-                        )
-                        .as_str(),
-                    )
-                });
-        }
+        require!(matches!(self.client_mode, ClientMode::SubmitHeader));
 
-        let submitter = env::predecessor_account_id();
-        self.update_submitter(&submitter, 1);
         let block_hash = block_header.calculate_hash();
-        #[cfg(feature = "logs")]
-        env::log_str(
+        require!(
+            block_hash
+                == self
+                    .unfinalized_tail_execution_header
+                    .as_ref()
+                    .map(|header| header.parent_hash)
+                    .unwrap_or(self.finalized_beacon_header.execution_block_hash),
             format!(
-                "Submitted header number {}, hash {:?}",
-                block_header.number, block_hash
+                "The block {:#?} has unknown parent {:#?}. Parent should be submitted first.",
+                block_hash, block_header.parent_hash
             )
-            .as_str(),
         );
 
-        let block_info = ExecutionHeaderInfo {
-            parent_hash: block_header.parent_hash,
-            block_number: block_header.number,
-            submitter,
-        };
-        let insert_result = self.unfinalized_headers.insert(&block_hash, &block_info);
+        let insert_result = self
+            .finalized_execution_blocks
+            .insert(&block_header.number, &block_hash);
+
         require!(
             insert_result.is_none(),
-            format!("The block {} already submitted!", &block_hash)
+            format!("The block {:#?} already submitted!", &block_hash)
+        );
+
+        let finalized_execution_header = self.finalized_execution_header.get().unwrap();
+        // Apply gc
+        if let Some(diff_between_unfinalized_head_and_tail) =
+            self.get_diff_between_unfinalized_head_and_tail()
```

Comment:
why it is important that it is not None? Maybe better use `unwrap_or(0)`?

Reply:
It is not too important, just to avoid calling the same code twice with `0` on the first and second calls in this mode.